### PR TITLE
New event system

### PR DESCRIPTION
**Pull request by kelunik**

### CODE REVIEW

File: src/Connection/Internal/Http1Parser.php [hunk 1]
```diff
@@ -116,9 +116,9 @@ public function parse(string $data = null): ?Response
         if (!$this->bodyStarted && \in_array($this->state, [self::BODY_CHUNKS, self::BODY_IDENTITY, self::BODY_IDENTITY_EOF], true)) {
             $this->bodyStarted = true;
 
+            $eventInvoker = events();
             \assert($this->response !== null);
-
-            events()->responseBodyStart($this->request, $this->stream, $this->response);
+            $eventInvoker->responseBodyStart($this->request, $this->stream, $this->response);
```
Comment (kelunik) [inline src/Connection/Internal/Http1Parser.php]: Why this change?

Comment (trowski) [reply src/Connection/Internal/Http1Parser.php]: The Psalm setting `rememberPropertyAssignmentsAfterCall="false"` means calling `events` loses the assertion that `$this->response !== null`

Comment (kelunik) [reply src/Connection/Internal/Http1Parser.php]: Shouldn't we store `$this->response` in a variable then instead of the event invoker?

File: composer-require-check.json [hunk 2]
```diff
@@ -16,6 +16,9 @@
         "void",
         "object",
         "mixed",
+        "random_bytes",
+        "events",
+        "processRequest",
```
Comment (kelunik) [inline composer-require-check.json]: Those are double present now, see below.

Comment (trowski) [reply composer-require-check.json]: Oops, automerge duplicated that since we put them in different places. Fixed.

File: src/Connection/Internal/Http1Parser.php (hunk 1 above)
Comment (trowski) [reply src/Connection/Internal/Http1Parser.php]: Either way, but that does match the style elsewhere better. Updated.

File: src/Connection/Internal/Http2ConnectionProcessor.php [hunk 3]
```diff
@@ -732,7 +717,7 @@ public function handleStreamException(Http2StreamException $exception): void
         $exception = new SocketException($exception->getMessage(), $code, $exception);
 
         if ($code === Http2Parser::REFUSED_STREAM) {
-            $exception = new UnprocessedRequestException($exception);
+            $exception = new UnprocessedRequestException($exception); // TODO
```
Comment (trowski) [inline src/Connection/Internal/Http2ConnectionProcessor.php]: There's still a TODO here, something missing yet?

File: src/Connection/Internal/Http2ConnectionProcessor.php [hunk 4]
```diff
@@ -1441,7 +1413,7 @@ private function shutdown(?HttpException $reason = null, ?int $lastId = null): v
             if ($lastId !== null && $id > $lastId) {
                 $exception = $exception instanceof UnprocessedRequestException
                     ? $exception
-                    : new UnprocessedRequestException($reason);
+                    : new UnprocessedRequestException($reason); // TODO
```
Comment (trowski) [inline src/Connection/Internal/Http2ConnectionProcessor.php]: Same here.

File: src/Connection/Internal/Http2ConnectionProcessor.php [hunk 5]
```diff
@@ -1035,19 +1017,18 @@ public function request(Request $request, Cancellation $cancellation, Stream $st
                     $writeFuture->await($cancellation);
 
                     if ($chunk === null) {
-                        $http2stream->requestBodyCompletion->complete();
+                        $http2stream->requestBodyCompletion->complete(); // TODO Move after loop?
```
Comment (trowski) [inline src/Connection/Internal/Http2ConnectionProcessor.php]: I think moving this after loop would be better to be triggered at a similar time to the event listener.

File: src/Connection/Internal/Http2ConnectionProcessor.php (hunk 3 above)
Comment (kelunik) [reply src/Connection/Internal/Http2ConnectionProcessor.php]: The branch gets rid of `UnprocessedRequestException` mostly, but not entirely, yet. Not sure why I mixed those changes into one branch.
